src/pf_driver/src/ros/pf_data_publisher.cpp

```diff
@@ -54,45 +54,92 @@ bool PFDataPublisher::stop()
   return true;
 }
 
+/** Update statistics for timesync based on packet reception time,
+ *  given a packet with sensor time and packet reception time. The
+ *  time_increment (sample period) must be known and passed in msg.
+ */
+template <typename T>
+void PFDataPublisher::update_timesync(T& packet)
+{
+  /* The timestamp in the packet tells about the time when the first sample in the
+     packet was measured, but now is rather shortly after the *last* sample in the
+     packet was measured (plus transmission and OS processing time that we don't know). */
+
+  /* To compute this, the time_increment must be known and up to date */
+  const auto time_for_points_in_packet =
+      rclcpp::Duration(0, packet.header.num_points_packet * (unsigned int)(1.0E9 * msg_->time_increment));
+  params_->passive_timesync.update(packet.header.timestamp_raw, 0, rclcpp::Clock().now() - time_for_points_in_packet);
+
+  RCLCPP_DEBUG(rclcpp::get_logger("timesync"), "packet#%d.%d.%d with %x %lu %.3f %u %u %f", scan_number_,
+               packet.header.header.scan_number, packet.header.header.packet_number, packet.header.status_flags,
+               packet.header.timestamp_raw, (rclcpp::Clock().now() - time_for_points_in_packet).seconds(),
+               packet.header.num_points_packet, packet.header.scan_frequency, 1.0E6 * msg_->time_increment);
+}
+
 // What are validation checks required here?
 // Skipped scans?
 // Device errors?
 template <typename T>
 void PFDataPublisher::to_msg_queue(T& packet, uint16_t layer_idx, int layer_inclination)
 {
-  if (!check_status(packet.header.status_flags))
-    return;
-
-  sensor_msgs::msg::LaserScan::SharedPtr msg;
-  if (d_queue_.empty())
-    d_queue_.emplace_back();
-  else if (d_queue_.size() > 5)
-    d_queue_.pop_front();
-  if (packet.header.header.packet_number == 1)
+  if (packet.header.status_flags != 0)
+  {
+    /* Information in packet is inconsistent.
+
+      Most probably (at startup or after changing parameters) with R2000 the
+      header.scan_frequency does not match the actual physical frequency ("unstable
+      rotation") and thus the time_increment from header does not match the
+      real sample frequency. TBD: Just ignore this packet or all up to now?
+    */
+    params_->passive_timesync.reset(0.0);
+  }
+
+  if (!!msg_ && (packet.header.header.scan_number == scan_number_))
```

Comment:
I think this is better readable and in addition with init of shared pointer etc. in class (see other suggestions).
```suggestion
  if ((msg_ != nullptr) && (packet.header.header.scan_number == scan_number_))
```